meson.build

```diff
@@ -148,6 +148,14 @@ if host_machine.system().endswith('bsd')
         deps_libpistache += libexecinfo_dep
 endif
 
+# libdl may be required for function dladdr, used in logStackTrace,
+# which is called by PS_LogWoBreak, used in turn by the stack-trace
+# logging macro PS_LOG_WO_BREAK_LIMITED and its derivatives. Issue
+# #1230.
+# Note: If 'dl' is not available, per Meson it suggests that the
+# functionality is provided by libc
+deps_libpistache += dependency('dl', required: false)
```

Comment:
Note that this will only work when users use Meson 0.62.0 or newer, while we only require Meson 0.53

Reply:
This is because `dependency()` in Meson can do different things when looking for a dependency "foo":

- search for a "foo" dependency overridden with `meson.override_dependency('foo', foo_dep)`
- search for a `foo.pc` file and use it
- search for `FooConfig.cmake` files and use them
- use a custom dependency resolution method if defined for the "foo" dependency

In our case:

- nobody calls `meson.override_dependency('dl')`
- no `dl.pc` file exists
- of course no `DlConfig.cmake` files exist either
- Meson defines a custom dependency resolution method for "dl", but this was added only in Meson 0.62.0

Until we'll require 0.62.0, we should do this instead:

```meson
deps_libpistache += compiler.find_library('dl', required: false)
```

And when we can require 0.62.0,

```meson
deps_libpistache += dependency('dl')
```

(note how `required: false` is absent, since we always require libdl's functionality. Meson will take care of figuring out when dl is provided by libc and won't fail when libdl doesn't exist)

Reply:
See the Meson docs:

- <https://mesonbuild.com/Reference-manual_functions.html#dependency>
- <https://mesonbuild.com/Dependencies.html#dependencies-with-custom-lookup-functionality>

Reply:
Hi -

Although I agree there is a counter argument, I was reluctant to dispense with the "required: false" for 
`deps_libpistache += dependency('dl')`
for fear meson would demand the presence of libdl in cases where we don't need it for real - noting that i) we almost never need it actually, and ii) we use only one potential-libdl function, and meson has no way of knowing which function we need (AFAIK) when determining if we need libdl. In other words, having "required: false" seems the more cautious path vs. breaking a build that might otherwise work.

Regarding meson version, could we / should we condition on the meson version like:
```
if meson.version().version_compare('>=0.62.0')
        deps_libpistache += dependency('dl', required: false)
else
    	deps_libpistache += compiler.find_library('dl', required: false)
endif
```
WDYT?

Reply:
> Although I agree there is a counter argument, I was reluctant to dispense with the "required: false" for `deps_libpistache += dependency('dl')` for fear meson would demand the presence of libdl in cases where we don't need it for real - noting that i) we almost never need it actually

We can find a solution which always works, see below :)

> , and ii) we use only one potential-libdl function, and meson has no way of knowing which function we need (AFAIK) when determining if we need libdl.

Of course it has! You can check if you have the potential-libdl function (which is it?) without linking to libdl, and then add libdl if needed:

```meson
has_func = compiler.has_function('potential_libdl')
if not has_func
  libdl_dep += dependency('dl')
  has_func = compiler.has_function('potential_libdl', dependencies: libdl_dep)
  if not has_func
    error('Unable to find potential_libdl(), even when linking to libdl')
  endif
  libpistache_deps += libdl_dep
endif
```

> In other words, having "required: false" seems the more cautious path vs. breaking a build that might otherwise work.
> 
> Regarding meson version, could we / should we condition on the meson version like:
> 
> ```
> if meson.version().version_compare('>=0.62.0')
>         deps_libpistache += dependency('dl', required: false)
> else
>     	deps_libpistache += compiler.find_library('dl', required: false)
> endif
> ```
> 
> WDYT?

The only reason why `dependency('dl')` exists is to be able to always require that dl functionality is available even when it is part of libc and no separate libdl exists. If we don't always require dl functionality, we might as well always use `compiler.find_library('dl', required: false)` instead.

In other words:

| code | meaning |
|---|---|
| `dependency('dl')` | always require dl functionality, whether it's part of libc or in a separate libdl object |
| `compiler.find_library('dl')` | always look for a separate libdl object |
| `compiler.find_library('dl', required: false)` | link to libdl, only if found |
| `dependency('dl', required: false)` | same as above, but requires meson 0.62.0 |